File: backend/elixir/omegle_phoenix/lib/omegle_phoenix/redis.ex
# Pairline - Open Source Video Chat and Matchmaking
# Copyright (C) 2026 Albert Blasczykowski
# Aless Microsystems
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

defmodule OmeglePhoenix.Redis do
  use Supervisor
  require OpenTelemetry.Tracer, as: Tracer
  alias OmeglePhoenix.Tracing

  alias OmeglePhoenix.Redis.AdminSubscriber

  @cluster_name :omegle_phoenix_redis_cluster
  @default_timeout 60_000
  @any_route_key "__eredis_cluster_any__"
  @any_commands MapSet.new([
                  "CLUSTER",
                  "COMMAND",
                  "DBSIZE",
                  "INFO",
                  "PING",
                  "PUBLISH",
                  "PUBSUB"
                ])
  @integer_response_commands MapSet.new([
                               "DEL",
                               "DBSIZE",
                               "EXISTS",
                               "EXPIRE",
                               "PEXPIRE",
                               "PUBLISH",
                               "SADD",
                               "SCARD",
                               "SREM",
                               "XACK",
                               "ZADD",
                               "ZCARD",
                               "ZREM"
                             ])

  def start_link(opts \\ []) do
    Supervisor.start_link(__MODULE__, opts, name: __MODULE__)
  end

  def command(cmd, opts \\ []) do
    command = normalize_command(cmd)
    command_name = command_name(command)

    Tracer.with_span "redis.command", %{kind: :client} do
      Tracing.annotate_client("redis.command")

      Tracer.set_attributes(%{
        "db.system" => "redis",
        "db.operation" => command_name,
        "redis.args_count" => max(length(command) - 1, 0)
      })

      with_timeout(opts, fn ->
        case :eredis_cluster.q(@cluster_name, command) do
          {:error, :invalid_cluster_command} ->
            :eredis_cluster.qk(@cluster_name, command, @any_route_key)

          other ->
            other
        end
      end)
      |> normalize_command_result(command_name)
      |> tap_redis_result()
    end
  end

  def pipeline(commands, opts \\ [])
  def pipeline([], _opts), do: {:ok, []}

  def pipeline(commands, opts) do
    normalized = Enum.map(commands, &normalize_command/1)

    Tracer.with_span "redis.pipeline", %{kind: :client} do
      Tracing.annotate_client("redis.pipeline")

      Tracer.set_attributes(%{
        "db.system" => "redis",
        "redis.pipeline.size" => length(normalized)
      })

      with_timeout(opts, fn -> run_pipeline(normalized) end)
      |> normalize_pipeline_result(normalized)
      |> tap_redis_result()
    end
  end

  def mget(keys, opts \\ [])

  def mget([], _opts), do: {:ok, []}

  def mget(keys, opts) when is_list(keys) do
    Tracer.with_span "redis.mget", %{kind: :client} do
      Tracing.annotate_client("redis.mget")

      Tracer.set_attributes(%{
        "db.system" => "redis",
        "db.operation" => "MGET",
        "redis.key_count" => length(keys)
      })

      keys
      |> Enum.map(&["GET", &1])
      |> pipeline(opts)
      |> tap_redis_result()
    end
  end

  def publish(channel, message, opts \\ []) do
    payload = JSON.encode!(message)

    Tracer.with_span "redis.publish", %{kind: :client} do
      Tracing.annotate_client("redis.publish")

      Tracer.set_attributes(%{
        "db.system" => "redis",
        "db.operation" => "PUBLISH",
        "messaging.destination" => channel
      })

      command(["PUBLISH", channel, payload], opts)
      |> tap_redis_result()
    end
  end

  def subscribe(_channel), do: {:error, :unsupported}
  def unsubscribe(_channel), do: {:error, :unsupported}

  @impl true
  def init(_opts) do
    with :ok <- ensure_cluster_started(),
         :ok <- connect_cluster() do
      Supervisor.init([AdminSubscriber], strategy: :one_for_one)
    else
      {:error, reason} ->
        {:stop, {:redis_cluster_connect_failed, reason}}
    end
  end

  defp ensure_cluster_started do
    case Application.ensure_all_started(:eredis_cluster) do
      {:ok, _apps} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  defp connect_cluster do
    init_nodes =
      redis_init_nodes()
      |> Enum.map(fn {host, port} -> {String.to_charlist(host), port} end)

    Application.put_env(:eredis_cluster, :pool_size, OmeglePhoenix.Config.get_redis_pool_size())

    options =
      case OmeglePhoenix.Config.get_redis_password() do
        nil -> []
        password -> [{:password, String.to_charlist(password)}]
      end

    disconnect_cluster()

    case :eredis_cluster.connect(@cluster_name, init_nodes, options) do
      :ok -> :ok
      {:error, {:already_started, _pid}} -> :ok
      {:error, reason} -> {:error, reason}
      other -> {:error, other}
    end
  end

  defp redis_init_nodes do
    case OmeglePhoenix.Config.get_redis_cluster_nodes() do
      [] -> [{OmeglePhoenix.Config.get_redis_host(), OmeglePhoenix.Config.get_redis_port()}]
      nodes -> nodes
    end
  end

  defp disconnect_cluster do
    try do
      :eredis_cluster.disconnect(@cluster_name)
    catch
      _, _ -> :ok
    end
  end

  defp tap_redis_result({:error, reason} = result) do
    Tracer.set_attribute("redis.outcome", "error")
    Tracer.set_attribute("error", true)
    Tracer.set_attribute("error.type", inspect(reason))
    result
  end

  defp tap_redis_result(result) do
    Tracer.set_attribute("redis.outcome", "ok")
    result
  end

  defp with_timeout(opts, fun) do
    timeout = Keyword.get(opts, :timeout, @default_timeout)

    timer_ref = Process.send_after(self(), {:redis_timeout, make_ref()}, timeout)

    try do
      fun.()
    catch
      :exit, {:timeout, _} -> {:error, :timeout}
    after
      Process.cancel_timer(timer_ref)
    end
  end

  defp run_pipeline(commands) do
    cond do
      Enum.all?(commands, &any_command?/1) ->
        :eredis_cluster.qk(@cluster_name, commands, @any_route_key)

      Enum.any?(commands, &any_command?/1) ->
        run_mixed_pipeline(commands)

      true ->
        :eredis_cluster.qmn(@cluster_name, commands)
    end
  end

  defp run_mixed_pipeline(commands) do
    {keyless_entries, keyed_entries} =
      commands
      |> Enum.with_index()
      |> Enum.split_with(fn {command, _index} -> any_command?(command) end)

    with {:ok, keyless_results} <- run_indexed_pipeline(keyless_entries, :keyless),
         {:ok, keyed_results} <- run_indexed_pipeline(keyed_entries, :keyed) do
      keyless_results
      |> Kernel.++(keyed_results)
      |> Enum.sort_by(&elem(&1, 0))
      |> Enum.map(&elem(&1, 1))
    end
  end

  defp run_indexed_pipeline([], _kind), do: {:ok, []}

  defp run_indexed_pipeline(entries, :keyless) do
    commands = Enum.map(entries, &elem(&1, 0))

    case :eredis_cluster.qk(@cluster_name, commands, @any_route_key) do
      {:ok, results} when is_list(results) ->
        {:ok, Enum.zip(Enum.map(entries, &elem(&1, 1)), results)}

      results when is_list(results) ->
        {:ok, Enum.zip(Enum.map(entries, &elem(&1, 1)), results)}

      {:error, reason} ->
        {:error, reason}

      other ->
        {:error, {:unexpected_pipeline_result, other}}
    end
  end

  defp run_indexed_pipeline(entries, :keyed) do
    commands = Enum.map(entries, &elem(&1, 0))

    case :eredis_cluster.qmn(@cluster_name, commands) do
      {:ok, results} when is_list(results) ->
        {:ok, Enum.zip(Enum.map(entries, &elem(&1, 1)), results)}

      results when is_list(results) ->
        {:ok, Enum.zip(Enum.map(entries, &elem(&1, 1)), results)}

      {:error, reason} ->
        {:error, reason}

      other ->
        {:error, {:unexpected_pipeline_result, other}}
    end
  end

  defp normalize_pipeline_result(results, commands) when is_list(results) and is_list(commands) do
    Enum.zip(results, commands)
    |> Enum.reduce_while({:ok, []}, fn
      {{:ok, value}, command}, {:ok, acc} ->
        {:cont, {:ok, [normalize_redis_value(command_name(command), value) | acc]}}

      {{:error, reason}, _command}, _acc ->
        {:halt, {:error, reason}}

      {other, _command}, _acc ->
        {:halt, {:error, {:unexpected_pipeline_entry, other}}}
    end)
    |> case do
      {:ok, values} -> {:ok, Enum.reverse(values)}
      error -> error
    end
  end

  defp normalize_pipeline_result({:ok, results}, commands) when is_list(results),
    do: normalize_pipeline_result(results, commands)

  defp normalize_pipeline_result({:error, reason}, _commands), do: {:error, reason}

  defp normalize_pipeline_result(other, _commands),
    do: {:error, {:unexpected_pipeline_result, other}}

  defp normalize_command_result({:ok, value}, command_name),
    do: {:ok, normalize_redis_value(command_name, value)}

  defp normalize_command_result({:error, reason}, _command_name), do: {:error, reason}
  defp normalize_command_result(other, _command_name), do: other

  defp normalize_command(command) when is_list(command) do
    Enum.map(command, &normalize_value/1)
  end

  defp command_name([command | _rest]), do: String.upcase(to_string(command))
  defp command_name(_command), do: nil

  defp any_command?(command), do: MapSet.member?(@any_commands, command_name(command))

  defp normalize_value(value) when is_binary(value), do: value
  defp normalize_value(value) when is_atom(value), do: Atom.to_string(value)
  defp normalize_value(value) when is_integer(value), do: Integer.to_string(value)

  defp normalize_value(value) when is_float(value),
    do: :erlang.float_to_binary(value, [:compact])

  defp normalize_value(value), do: to_string(value)

  defp normalize_redis_value(_command_name, :undefined), do: nil

  defp normalize_redis_value(command_name, value) when is_binary(value) do
    if MapSet.member?(@integer_response_commands, command_name) do
      case Integer.parse(value) do
        {parsed, ""} -> parsed
        _ -> value
      end
    else
      value
    end
  end

  defp normalize_redis_value(command_name, value) when is_list(value) do
    Enum.map(value, &normalize_redis_value(command_name, &1))
  end

  defp normalize_redis_value(command_name, value) when is_tuple(value) do
    value
    |> Tuple.to_list()
    |> Enum.map(&normalize_redis_value(command_name, &1))
    |> List.to_tuple()
  end

  defp normalize_redis_value(_command_name, value), do: value
end

File: backend/golang/internal/redis/redis.go
// Pairline - Open Source Video Chat and Matchmaking
// Copyright (C) 2026 Albert Blasczykowski
// Aless Microsystems
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package redis

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	redisotel "github.com/redis/go-redis/extra/redisotel/v9"
	"github.com/redis/go-redis/v9"
	"github.com/redis/go-redis/v9/maintnotifications"
)

const AdminActionStream = "admin:action:stream"

type Client struct {
	client redis.UniversalClient
}

func NewClient() *Client {
	addrs, err := redisClusterAddrsFromEnv()
	if err != nil {
		log.Fatal(err)
	}

	password := os.Getenv("REDIS_PASSWORD")

	rdb := redis.NewClusterClient(&redis.ClusterOptions{
		Addrs:                    addrs,
		Password:                 password,
		MaintNotificationsConfig: redisMaintNotificationsConfig(),
	})

	if err := redisotel.InstrumentTracing(rdb); err != nil {
		log.Printf("Failed to enable Redis tracing: %v", err)
	}

	if err := redisotel.InstrumentMetrics(rdb); err != nil {
		log.Printf("Failed to enable Redis metrics: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := rdb.Ping(ctx).Err(); err != nil {
		_ = rdb.Close()
		log.Fatalf("Failed to connect to Redis Cluster: %v", err)
	}

	log.Println("Connected to Redis Cluster successfully")
	return &Client{client: rdb}
}

func redisMaintNotificationsConfig() *maintnotifications.Config {
	raw := strings.TrimSpace(strings.ToLower(os.Getenv("REDIS_MAINT_NOTIFICATIONS_MODE")))

	switch raw {
	case "", "disabled", "false", "0":
		return &maintnotifications.Config{Mode: maintnotifications.ModeDisabled}
	case "auto":
		return &maintnotifications.Config{Mode: maintnotifications.ModeAuto}
	case "enabled", "true", "1":
		return &maintnotifications.Config{Mode: maintnotifications.ModeEnabled}
	default:
		log.Printf("Unknown REDIS_MAINT_NOTIFICATIONS_MODE=%q; defaulting to disabled", raw)
		return &maintnotifications.Config{Mode: maintnotifications.ModeDisabled}
	}
}

func (r *Client) PublishBanAction(ctx context.Context, sessionID, ipAddress, reason string) error {
	data := map[string]interface{}{
		"action":     "emergency_ban",
		"session_id": sessionID,
		"ip":         ipAddress,
		"reason":     reason,
		"timestamp":  time.Now().UnixMilli(),
	}

	return r.publishJSON(ctx, data)
}

func (r *Client) PublishBanIPAction(ctx context.Context, ipAddress, reason string) error {
	data := map[string]interface{}{
		"action":    "emergency_ban_ip",
		"ip":        ipAddress,
		"reason":    reason,
		"timestamp": time.Now().UnixMilli(),
	}

	return r.publishJSON(ctx, data)
}

func (r *Client) PublishUnbanAction(ctx context.Context, sessionID, ipAddress string) error {
	data := map[string]interface{}{
		"action":     "emergency_unban",
		"session_id": sessionID,
		"ip":         ipAddress,
		"timestamp":  time.Now().UnixMilli(),
	}

	return r.publishJSON(ctx, data)
}

func (r *Client) PublishUnbanIPAction(ctx context.Context, ipAddress string) error {
	data := map[string]interface{}{
		"action":    "emergency_unban_ip",
		"ip":        ipAddress,
		"timestamp": time.Now().UnixMilli(),
	}

	return r.publishJSON(ctx, data)
}

func (r *Client) PublishDisconnectAction(ctx context.Context, sessionID string) error {
	data := map[string]interface{}{
		"action":     "emergency_disconnect",
		"session_id": sessionID,
		"timestamp":  time.Now().UnixMilli(),
	}

	return r.publishJSON(ctx, data)
}

func (r *Client) publishJSON(ctx context.Context, data map[string]interface{}) error {
	payload, err := json.Marshal(data)
	if err != nil {
		return err
	}

	err = r.client.XAdd(ctx, &redis.XAddArgs{
		Stream: AdminActionStream,
		MaxLen: 10_000,
		Approx: true,
		Values: map[string]interface{}{
			"payload": string(payload),
		},
	}).Err()
	return err
}

func (r *Client) Close() error {
	return r.client.Close()
}

func SetIndexedValue(
	ctx context.Context,
	client redis.UniversalClient,
	indexKey,
	key,
	value string,
	ttl time.Duration,
) error {
	if err := client.Set(ctx, key, value, ttl).Err(); err != nil {
		return err
	}

	if err := client.SAdd(ctx, indexKey, key).Err(); err != nil {
		if rollbackErr := client.Del(ctx, key).Err(); rollbackErr != nil {
			return fmt.Errorf("add index entry: %w (rollback delete failed: %v)", err, rollbackErr)
		}

		return fmt.Errorf("add index entry: %w", err)
	}

	return nil
}

func DeleteIndexedKey(ctx context.Context, client redis.UniversalClient, indexKey, key string) error {
	if err := client.Del(ctx, key).Err(); err != nil {
		return err
	}

	if err := client.SRem(ctx, indexKey, key).Err(); err != nil {
		return fmt.Errorf("remove index entry: %w", err)
	}

	return nil
}

// GetClient returns the underlying redis client
func (r *Client) GetClient() redis.UniversalClient {
	return r.client
}

func redisClusterAddrsFromEnv() ([]string, error) {
	clusterNodes := strings.TrimSpace(os.Getenv("REDIS_CLUSTER_NODES"))
	if clusterNodes == "" {
		return nil, fmt.Errorf("REDIS_CLUSTER_NODES is required for Go services")
	}

	parts := strings.Split(clusterNodes, ",")
	addrs := make([]string, 0, len(parts))

	for _, part := range parts {
		addr := strings.TrimSpace(part)
		if addr != "" {
			addrs = append(addrs, addr)
		}
	}

	if len(addrs) == 0 {
		return nil, fmt.Errorf("REDIS_CLUSTER_NODES must contain at least one host:port entry")
	}

	return addrs, nil
}
